### PR TITLE
initial sparql mapping commit

### PR DESCRIPTION
**Pull request by bdevloed**
Adds sparql construct based mapping.

### CODE REVIEW

**Comment by bdevloed** [comment]
Still being tested -  PR in draft

File: README.md [hunk 1]
```diff
@@ -30,6 +30,40 @@ The context is configured through `delta-context-config.js` which is covered in
...
 
+### Delta Message - SPARQL mapping
+
+The delta-consumer provides a way to map the incoming messages to a different model. This is done by providing a SPARQL queries in the configuration directory. The queries are executed on the landing zone graph and the results are used to update target graph.
+
+Each triple from the delta message is treated individually.
+
+#### Delete
+
+When a delete comes in, which breaks the where part of a query - the whole construct clause is deleted from the target graph.
```
Comment (cecemel) [inline README.md:41]: I remember we discovered some kind of 'unexpected' behavior; when the following mapping would be done:

Suppose incoming delta:
```
DELETE <subject> skos:inScheme <http://foo>
```

and mapping:
```
CONSTRUCT {
   ?s skos:prefLabel ?label.
} WHERE {
   ?s skos:inScheme <http://foo>;
     skos:prefLabel ?label.
}
```

would result in:
```
DELETE <subject> skos:prefLabel "label"
```

which might feel unexpected; but I remember we talked a fair bit about it; and I think we ended with what has been described in the `README.md`. But I can't find it in my notes, so maybe, is that what I am looking for: a wrap-up of the discussion about these 'unexpected' deltas? Do you have it somewhere?

File: lib/delta-sparql-mapping.js [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+    console.log(
+      `Successfully executed initial mapping query (${insertQueries.indexOf(query) + 1} of ${insertQueries.length})`,
+    );
+  }
+  const endTime = Date.now();
+  console.log(`Initial mapping took ${endTime - startTime}ms`);
+}
+
+export async function synchronizedDeltaProcessing(deltaMessage) {
+  await lock.acquire();
```
Comment (cecemel) [inline lib/delta-sparql-mapping.js:385]: I am not sure a lock is needed here.
Looking at the code in `delta-sync.js`, it indeed looks like it's possible for one delta-consumer job to start while another one is busy (and really doing something, not just due to a premature restart). And it makes sense not to want that. However, maybe I would fix it there, and not in the core logic of the `delta-sparql-mapping.js`. What do you think?

File: pipelines/delta-sync.js [hunk 3]
```diff
@@ -74,12 +84,21 @@ async function runDeltaSync() {
...
-            const { termObjectChangeSets, termObjectChangeSetsWithContext } = await deltaFile.load();
-            console.log(`
-              Dispatching ${termObjectChangeSets.length} term object change sets`
-              + ` and ${termObjectChangeSetsWithContext.length} term object change sets with context
-            `);
-            await deltaSyncDispatching.dispatch({ mu, muAuthSudo, fetch, chunk, sparqlEscapeUri: mu.sparqlEscapeUri }, { termObjectChangeSets, termObjectChangeSetsWithContext }, constants);
+          } else if (ENABLE_SPARQL_MAPPING) {
+            const { termObjectChangeSets, changeSets } = await deltaFile.load();
+            await synchronizedDeltaProcessing(changeSets);
+            // To be discussed: keep the custom dispatching when mapping is enabled
```
Comment (cecemel) [inline pipelines/delta-sync.js:90]: Why wouldn't we want to keep it?

**Review by cecemel — COMMENTED** [review]
I started looking into the code; maybe already a few comments. Not finished yet

File: README.md [hunk 4]
```diff
@@ -145,8 +266,46 @@ When adding rules and queries to the reasoner, make sure the required context is
...
+      DCR_DISABLE_DELTA_INGEST: "false"
+      DCR_DISABLE_INITIAL_SYNC: "false"
+      DCR_WAIT_FOR_INITIAL_SYNC: "true"
+      BYPASS_MU_AUTH_FOR_EXPENSIVE_QUERIES: "true"
+      DCR_TARGET_DATABASE: "virtuoso"
+      DCR_ENABLE_SPARQL_MAPPING: "true"
+      DCR_LANDING_ZONE_GRAPH: "http://mu.semte.ch/graphs/op-consumer-test"
+      DCR_TARGET_GRAPH: "http://mu.semte.ch/graphs/op-consumer-test-transformed"
+    volumes:
+      - .:/app/
```
Comment (cecemel) [inline README.md:304]: maybe not needed?

File: README.md [hunk 5]
```diff
@@ -145,8 +266,46 @@ When adding rules and queries to the reasoner, make sure the required context is
...
+      - ./config/:/config/
```
Comment (cecemel) [inline README.md:305]: this one too?

**Comment by cecemel** [comment]
Don't merge yet, found a regression in the single-graph-consumer logic

**Comment by cecemel** [comment]
ok should be fixed now

**Pull request merged**